sql/parser/eval.go

```diff
@@ -1144,3 +1157,82 @@ func ParseTimestamp(s DString) (DTimestamp, error) {
 	// Parse other formats in the future.
 	return DummyTimestamp, err
 }
+
+// SimilarEscape converts a SQL:2008 regexp pattern to POSIX style, so it can
+// be used by our regexp engine.
+func SimilarEscape(pattern string) string {
+	return similarEscapeCustomChar(pattern, '\\')
+}
+
+// similarEscapeCustomChar converts a SQL:2008 regexp pattern to POSIX style,
+// so it can be used by our regexp engine. This version of the function allows
+// for a custom escape character.
+func similarEscapeCustomChar(pattern string, escapeChar rune) string {
+	patternBuilder := make([]rune, 0, utf8.RuneCountInString(pattern))
+
+	inCharClass := false
+	afterEscape := false
+	numQuotes := 0
+	for _, c := range pattern {
+		switch {
+		case afterEscape:
+			// For SUBSTRING patterns
+			if c == '"' && !inCharClass {
+				if numQuotes%2 == 0 {
+					patternBuilder = append(patternBuilder, '(')
+				} else {
+					patternBuilder = append(patternBuilder, ')')
+				}
+				numQuotes++
+			} else {
+				patternBuilder = append(patternBuilder, '\\', c)
+			}
+			afterEscape = false
+		case utf8.ValidRune(escapeChar) && c == escapeChar:
+			// SQL99 escape character; do not immediately send to output
+			afterEscape = true
+		case inCharClass:
+			if c == '\\' {
+				patternBuilder = append(patternBuilder, '\\')
+			}
+			patternBuilder = append(patternBuilder, c)
+			if c == ']' {
+				inCharClass = false
+			}
+		case c == '[':
+			patternBuilder = append(patternBuilder, c)
+			inCharClass = true
+		case c == '%':
+			patternBuilder = append(patternBuilder, '.', '*')
+		case c == '_':
+			patternBuilder = append(patternBuilder, '.')
+		case c == '(':
+			// Convert to non-capturing parenthesis
+			patternBuilder = append(patternBuilder, '(', '?', ':')
+		case c == '\\', c == '.', c == '^', c == '$':
+			// Escape these characters because they are NOT
+			// metacharacters for SQL-style regexp
+			patternBuilder = append(patternBuilder, '\\', c)
+		default:
+			patternBuilder = append(patternBuilder, c)
+		}
+	}
+
+	return string(patternBuilder)
+}
+
+// wrapPattern surrounds the transformed input string with
+//   ^(?: ... )$
+// which requires some explanation.  We need "^" and "$" to force
+// the pattern to match the entire input string as per SQL99 spec.
+// The "(?:" and ")" are a non-capturing set of parens; we have to have
+// parens in case the string contains "|", else the "^" and "$" will
+// be bound into the first and last alternatives which is not what we
+// want, and the parens must be non capturing because we don't want them
+// to count when selecting output for SUBSTRING.
+func wrapPattern(pattern string, caseInsensitive bool) string {
```

Comment:
I'd name this `anchorPattern` or `anchorRegex` in order to be a bit more specific about its operation.